python_bindings/src/builtin_helpers_src.py

```diff
@@ -0,0 +1,787 @@
+            self._requirements[name]._check_types_and_dims(
+                f.output_types(), f.dimensions())
+            funcs.append(f)
+
+        self._pipeline = Pipeline(funcs)
+        self._stage = _Stage.PipelineBuilt
+        return self._pipeline
+
+    def _get_input_parameter(self, name: str) -> list[InternalParameter]:
+        assert self._stage == _Stage.PipelineBuilt
+        _check(name in self._input_parameters, "Unknown input: %s" % name)
+        return [self._input_parameters[name]]
+
+    def _get_output_func(self, name: str) -> list[Func]:
+        assert self._stage == _Stage.PipelineBuilt
+        _check(name in self._output_funcs, "Unknown output: %s" % name)
+        return [self._output_funcs[name]]
+
+    def _bind_input(self, name: str, values: list[object]):
+        assert (self._stage == _Stage.GeneratorParamsBuilt
+                or self._stage == _Stage.IOBuilt)
+        _check(
+            len(values) == 1, "Too many values specified for input: %s" % name)
+        _check(name in self._inputs_names,
+               "There is no input with the name: %s" % name)
+        g = getattr(self._inputs, name)
+        assert not name in self._replacements
+        self._replacements[name] = g._make_replacement(
+            values[0], self._requirements[name])
+
+
+_python_generators: dict = {}
+
+
+def _get_python_generator_names() -> list[str]:
+    return _python_generators.keys()
+
+
+def _fqname(o):
+    k = o
+    m = k.__module__
+    q = k.__qualname__
+    if m == "__main__" or k == "builtins":
+        return q
+    return m + "." + q
+
+
+def _find_python_generator_class(name: str):
+    cls = _python_generators.get(name, None)
+    if not isclass(cls):
+        cls = None
+    return cls
+
+
+def alias(**kwargs):
+
+    def alias_impl(cls):
+        nonlocal kwargs
+        for k, v in kwargs.items():
+            _check_valid_name(k)
+            _check(hasattr(cls, "_halide_registered_name"),
+                   "@alias can only be used in conjunction with @generator.")
+            _check(not k in _python_generators,
+                   "The Generator name %s is already in use." % k)
+            _check(
+                type(v) is dict,
+                "The Generator alias %s specifies something other than a dict."
+                % k)
+            new_cls = type(k, (cls, ), {
+                "_halide_registered_name": k,
+                "_halide_alias_generator_params": v
+            })
+            _python_generators[k] = new_cls
+        return cls
+
+    return alias_impl
+
+
+def generator(name=""):
+    # This code relies on dicts preserving key-insertion order, which is only
+    # guaranteed for all Python implementations as of v3.7.
+    assert sys.version_info >= (
+        3, 7), "Halide Generators require Python 3.7 or later."
+
+    def generator_impl(cls):
+        nonlocal name
+        if not name:
+            name = _fqname(cls)
```

Comment:
Because of nonlocal, this writes to generator's frame, not to generator_impl's. I don't see a good reason for this.

Reply:
how about replacing with `n = name if name else _fqname(cls)`

Reply:
`n = name or _fqname(cls)` is equivalent, but it's also totally kosher to re-use `name` as in `name = name or _fqname(cls)`.